src/lib/OS/SyscallHandler.cc

```diff
@@ -1,15 +1,712 @@
+      // mprotect is not supported
+      // always return zero to indicate success
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {0ull}};
+      break;
+    }
+    case 235: {  // mbind
+      // mbind is not supported due to all binaries being single threaded.
+      // Always return zero to indicate success
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {0ull}};
+      break;
+    }
+    case 261: {  // prlimit64
+      // TODO: Functionality temporarily omitted as it is unused within
+      // workloads regions of interest and not required for their simulation
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {0ull}};
+      break;
+    }
+    case 278: {  // getrandom
+      // TODO: support flags argument
+
+      // seed random numbers
+      srand(clock());
+
+      // Write <buflen> random bytes to buf
+      uint64_t bufPtr = info.registerArguments[0].get<uint64_t>();
+      size_t buflen = info.registerArguments[1].get<size_t>();
+
+      char buf[buflen];
+      for (size_t i = 0; i < buflen; i++) {
+        buf[i] = (uint8_t)rand();
+      }
+
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {(uint64_t)buflen}};
+
+      stateChange.memoryAddresses.push_back({bufPtr, (uint8_t)buflen});
+      stateChange.memoryAddressValues.push_back(RegisterValue(buf, buflen));
+
+      break;
+    }
+    case 293: {  // rseq
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {0ull}};
+      break;
+    }
+
+    default:
+      break;
+  };
+
+  concludeSyscall(stateChange);
+}
+
+void SyscallHandler::readStringThen(char* buffer, uint64_t address,
+                                    int maxLength,
+                                    std::function<void(size_t length)> then,
+                                    int offset) {
+  if (maxLength <= 0) {
+    return then(offset);
+  }
+
+  // Get a string from the simulation memory and within the passed buffer
+  std::vector<char> data = memory_->getUntimedData(address + offset, maxLength);
+
+  for (int i = 0; i < data.size(); i++) {
+    buffer[i] = data[i];
+    // End of string; call onwards
+    if (buffer[i] == '\0') return then(i + 1);
+  }
+
+  // Reached max length; call onwards
+  return then(maxLength);
+}
+
+void SyscallHandler::readBufferThen(uint64_t ptr, uint64_t length,
+                                    std::function<void()> then) {
+  // If there's nothing to read, consider the read to be complete and call
+  // onwards
+  if (length == 0) {
+    return then();
+  }
+
+  // Get data from the simulation memory and read into dataBuffer_
+  std::vector<char> data = memory_->getUntimedData(ptr, length);
+  dataBuffer_.insert(dataBuffer_.end(), data.begin(), data.begin() + length);
+
+  // Read in data, call onwards
+  return then();
+}
+
+void SyscallHandler::readLinkAt(span<char> path) {
+  if (path.size() == PATH_MAX_LEN) {
+    // TODO: Handle PATH_MAX_LEN case
+    std::cout << "\n[SimEng:SyscallHandler] Path exceeds PATH_MAX_LEN"
```

Comment:
What is the need for the newline in the error message?

Reply:
old formatting from original exceptionHandler